src/amp-story-player/amp-story-player-impl.js

```diff
@@ -1160,28 +1145,44 @@ export class AmpStoryPlayer {
 
   /**
    * Returns a promise that makes sure current story gets loaded first before
-   * others.
+   * others. When this method is called with any story other than the current
+   * story, it will not resolve until the current story has finished loading.
    * @param {!StoryDef} story
    * @return {!Promise}
    * @private
    */
   currentStoryPromise_(story) {
     if (this.stories_[this.currentIdx_].storyContentLoaded) {
+      // Return a promise signifying that the current story has loaded.
       return Promise.resolve();
     }
 
     if (story.distance !== 0) {
+      // When the current story has not yet loaded and the given story is not
+      // the current story, return a promise that will resolve only when the
+      // current story has loaded.
       return this.currentStoryLoadDeferred_.promise;
     }
 
-    if (this.currentStoryLoadDeferred_) {
-      // Cancel previous story load promise.
-      this.currentStoryLoadDeferred_.reject(
-        `[${LOG_TYPE_ENUM.DEV}] Cancelling previous story load promise.`
-      );
-    }
+    // Cancel previous story load promise.
+    this.currentStoryLoadDeferred_?.reject(
+      `[${LOG_TYPE_ENUM.DEV}] Cancelling previous story load promise.`
+    );
+
+    // Reset `this.currentStoryLoadDeferred_` and have it resolve only when the
+    // current story has loaded.
+    this.currentStoryLoadDeferred_ = new Deferred();
+    story.messagingPromise.then((messaging) =>
+      messaging.registerHandler('storyContentLoaded', () => {
+        // Stories that already loaded won't dispatch a `storyContentLoaded`
+        // event anymore, which is why we need this sync property.
+        story.storyContentLoaded = true;
```

Comment:
If you want to keep the `onStoryContentLoaded_` helper we should use it here. Since it's just one line and not expected to grow at all, I'm OK with just repeating `story.storyContentLoaded = true;` in both listeners.